### PR TITLE
Load lua code from cell data

### PR DESCRIPTION
**Pull request by contrun**

### CODE REVIEW

File: lua-loader/lua-loader.c [hunk 1]
```diff
@@ -309,6 +384,10 @@ static int pmain(lua_State *L) {
     if (args & has_r) {
         if (!run_from_file(L)) return 0;
     }
+    // No arguments found, trying to load lua code from cell data
+    if (argc == 1) {
+        if (load_lua_code_from_cell_data(L)) return 0;
```
Comment (XuJiandong) [inline lua-loader/lua-loader.c:389]: Parse script args here then get `code_hash` and `hash_type`. Then use `load_lua_code_with_hash`

File: lua-loader/lua-loader.c [hunk 2]
```diff
@@ -143,6 +143,81 @@ static int dostring(lua_State *L, const char *s, const char *name) {
...
+                    goto eval;
+                }
+                break;
+            default:
+                return ret;
+        }
+        current++;
+    }
+eval:
+    return dostring(L, (const char *)buf, __func__);
```
Comment (XuJiandong) [inline lua-loader/lua-loader.c:218]: should make sure buf is ended with \0. Append one if necessary.

File: lua-loader/lua-loader.c [hunk 3]
```diff
@@ -143,6 +143,81 @@ static int dostring(lua_State *L, const char *s, const char *name) {
...
+            default:
+                return CKB_INDEX_OUT_OF_BOUND;
+        }
+        current++;
+    }
+eval:
+    return dostring(L, (const char *)buf, __func__);
+}
+
+int load_lua_code_with_hash(lua_State *L, const uint8_t *code_hash,
```
Comment (XuJiandong) [inline lua-loader/lua-loader.c:176]: should use `ckb_look_for_dep_with_hash2` then call `ckb_load_cell_data`.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: lua-loader/lua-ckb.c [hunk 4]
```diff
@@ -228,146 +233,143 @@ int CKB_LOAD_V5(lua_State *L, syscall_v5 f) {
 //         perLine: number of bytes on each output line.
 
 void hexDump(const char *desc, const void *addr, const int len, int perLine) {
```
[the remark this answers is not in the document]
Comment (contrun) [reply lua-loader/lua-ckb.c:238]: Will refactor it in a separate PR, which would also address the problem of overusing macros.

File: lua-loader/lua-loader.c [hunk 5]
```diff
@@ -309,6 +374,10 @@ static int pmain(lua_State *L) {
     if (args & has_r) {
         if (!run_from_file(L)) return 0;
     }
+    // No arguments found, trying to load lua code from cell data
+    if (argc == 1) {
```
Comment (XuJiandong) [inline lua-loader/lua-loader.c:378]: It should be `argc == 0` without ckb-debugger enviroment.

**Review by XuJiandong — APPROVED** [review]

File: lua-loader/lua-loader.c (hunk 5 above)
Comment (contrun) [reply lua-loader/lua-loader.c:378]: I don't want to touch it in this PR, as there are many other places we need to change, cf https://github.com/XuJiandong/ckb-lua/issues/25.

**Pull request merged**